src/ir/auth_logic/ast.h

```diff
@@ -0,0 +1,176 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+// This file describes AST nodes for authorization logic
+
+#ifndef SRC_IR_AUTH_LOGIC_AST_H_
+#define SRC_IR_AUTH_LOGIC_AST_H_
+
+#include <string>
+#include <variant>
+#include <vector>
+
+namespace raksha::ir::auth_logic {
+
+class Principal {
+ public:
+  Principal(std::string name) : name_(std::move(name)) {}
+
+ private:
+  std::string name_;
+};
+
+// Used to represent whether a predicate is negated or not
+enum Sign { kNegated, kPositive };
+
+class Predicate {
+ public:
+  Predicate(std::string name, std::vector<std::string> args, Sign sign)
+      : name_(std::move(name)),
+        args_(std::move(args)),
+        // TODO move surpresses an unused private field error about sign_
+        // for now, get rid of this eventually
+        sign_(std::move(sign)) {}
+
+ private:
+  std::string name_;
+  std::vector<std::string> args_;
+  Sign sign_;
+};
+
+// This represents an expression that applies either a predicate or a
+// "canActAs" to a principal, and it is a sub-term of FlatFact
+//    - pred_name(arg_1, ..., arg_n)           (a predicate)
+//    - principal_1 canActAs principal_2       (a canActAs)
+// To encapsulate which form this takes, it:
+//    - uses factory methods to make either form
+//    - uses a private std::variant member that takes either of these two forms
+//    - uses std::visit to do operations on the actual state
+class VerbPhrase {
+ public:
+  explicit VerbPhrase(Predicate pred) : internal_form_(std::move(pred)) {}
+  explicit VerbPhrase(Principal right_prin)
+      : internal_form_(std::move(right_prin)) {}
+
+ private:
+  // The principal variant of the internal form means this represents
+  // "canActAs principal".
+  std::variant<Predicate, Principal> internal_form_;
+};
+
+// This the AST Node for FlatFact which takes on one of two forms:
+//   - <principal> <verb_phrase>  (* a principal applied to a verbphrase *)
```

Comment:
What would this correspond to in the authorization logic syntax? I am asking because none of the following are valid syntax:
  - `"UserX" "UserY"`
  - `"UserX" claimHasTag("...")`

Reply:
we discussed by video